Dispatch Managing Actor

Marenz opened the pull request:
- **Reset release notes**
- **Add dispatch runner**

Marenz committed:
Send STOPPED message when duration is reached

Signed-off-by: Mathias L. Baumann <mathias.baumann@frequenz.com>

Marenz committed:
Add dispatch mananging actor

A useful actor to help control and manange another actor using
dispatches.

Signed-off-by: Mathias L. Baumann <mathias.baumann@frequenz.com>

daniel-zullo-frequenz commented on tests/test_runner.py:
```diff
@@ -0,0 +1,171 @@
+    await runner_actor.stop()
+
+
+# Disable for _handle_dispatch access
+# pylint: disable=protected-access
```
These lines might no longer be needed

daniel-zullo-frequenz commented on tests/test_runner.py:
```diff
@@ -0,0 +1,171 @@
+
+from frequenz.dispatch import Dispatch, DispatchConfigurationEvent, DispatchRunnerActor
+
+# pylint: disable=protected-access
```
Probably not needed

daniel-zullo-frequenz commented on src/frequenz/dispatch/_actor_runner.py:
```diff
@@ -0,0 +1,187 @@
+        Args:
+            dispatch: The dispatch to handle.
+
+        Returns:
+            The running state.
```
The function returns None :thinking:

daniel-zullo-frequenz reviewed (commented):
Just a few really minor comments to check for, LGTM otherwise

llucax commented on src/frequenz/dispatch/_actor_runner.py:
```diff
@@ -0,0 +1,187 @@
+    """Whether this is a dry run."""
+
+    payload: dict[str, Any]
+    """Additional payload."""
```
I'm still not 100% sold on this, I still feel that it might be better to send an actual `Dispatch` object, possibly a merged one (i.e. if we are merging all dispatches with different IDs but the same type, for example, and we have id=1 start=10:00 end=11:00 and id=2 start=11:00 end=None, that would be merged into a dispatch with start=10:00 and end=None).

But since that would be harder to implement I'm OK with starting with this simple approach and see if we need something more complicated in the future.

llucax replied:
* #55

llucax commented on src/frequenz/dispatch/_actor_runner.py:
```diff
@@ -0,0 +1,187 @@
+                    print("Dispatch: Setting components to %s", components)
+                case [ComponentCategory.BATTERY, *_]:
+                    print("Dispatch: Using all battery components")
+                case _ as unsupported:
```
I used this scheme until I realized that is exactly the same as `case unsupported:`.

llucax commented on src/frequenz/dispatch/_actor_runner.py:
```diff
@@ -0,0 +1,187 @@
+        )
+
+        # Create config channel to receive (re-)configuration events pre-start and mid-run
+        config_channel = Broadcast[DispatchConfigurationEvent](name="config_channel")
```
I know this is a separate repo, but I was confused for a second with the config manager. I would try to make a more explicit separation, maybe call this `dispatch_reconfiguration_channel` (or `dispatch_updates_channel`). I actually think `update` is a better name here, this is not about configuration, but about receiving updates to a running dispatch.

llucax commented on src/frequenz/dispatch/_actor_runner.py:
```diff
@@ -0,0 +1,187 @@
+
+        # Start actor and supporting actor, give each a config channel receiver
+        my_actor = MyActor(config_channel.new_receiver())
+        supporting_actor = MagicMock(config_channel.new_receiver())
```
I don't get what this `supporting_actor` is.

llucax commented on src/frequenz/dispatch/_actor_runner.py:
```diff
@@ -0,0 +1,187 @@
+
+        status_receiver = dispatcher.running_status_change.new_receiver()
+
+        dispatch_handler = DispatchRunnerActor(
```
I would call this `dispatch_runner = DispatchRunnerActor(...)` for consistency.

llucax commented on src/frequenz/dispatch/_actor_runner.py:
```diff
@@ -0,0 +1,187 @@
+
+    def __init__(
+        self,
+        actors: frozenset[Actor],
```
Do we have usecases for passing multiple actors to this actor?

Marenz replied:
Dunno, @daniel-zullo-frequenz seems to have required it, that part is from him :D

daniel-zullo-frequenz replied:
Yes, there is a case where we start/stop more than 2 actors. Is there any reason why you'd like to handle only a single actor?

llucax replied:
It feels like it is more natural to map one actor to handling one dispatch, but I'm not strongly opposed. I would like to hear about the details on your use case thought, maybe in private to avoid spamming the PR too much?

llucax replied:
After talking to @daniel-zullo-frequenz we agreed that in the future his use case should use only one central actor to handle dispatches that spawns more actors as needed, but that would mean a big refactoring we can start right now, so we would leave the support for multiple actors here for now. What we can do is accept `actor: Actor | Set[Actor]`, so for all normal use cases people can pass just one actor, and document the use of multiple actors is discouraged and will be removed in the future.

llucax replied:
Oh, I see you removed the support for multiple actors already. I didn't meant we should remove it, but we should *figure out if we really need to ...* :(

llucax commented on src/frequenz/dispatch/_actor_runner.py:
```diff
@@ -0,0 +1,187 @@
+        while True:
+            _logger.info("Waiting for dispatch...")
+            dispatch = await self._dispatch_rx.receive()
+            await self._handle_dispatch(dispatch=dispatch)
```
I would use the more idiomatic:

```py
        async for dispatch in self._dispatch_rx:
            await self._handle_dispatch(dispatch=dispatch)
```

And then add a log when a dispatch update is received (like "dispatch update received, processing...", probably with debug level) and finished processing (like "dispatch update processed"). You lose the explicit "waiting" info message, but it is also implicit that it is waiting once it finished processing a dispatch. You also get an automatic info log when the actor is started (via the `Actor` class).

llucax commented on src/frequenz/dispatch/_actor_runner.py:
```diff
@@ -0,0 +1,187 @@
+        match running:
+            case RunningState.STOPPED:
+                _logger.info("Stopping dispatch...")
+                await self._stop_actors("Dispatch stopped")
```
Maybe "Stopped by a dispatch".

llucax commented on src/frequenz/dispatch/_actor_runner.py:
```diff
@@ -0,0 +1,187 @@
+
+
+@dataclass(frozen=True, kw_only=True)
+class DispatchConfigurationEvent:
```
I would call this `DispatchUpdate`. I think calling this configuration is a bit confusing because we use *configuration* for something else, and also I think even should be "created", "updated", etc. For me this class only holds a *dispatch update*.

llucax reviewed (commented):
I would say all comments are mostly bike shedding, but in particular I think not using *configuration* and use *dispatch update* is important, I think it can get a bit confusing for people dealing with dispatches and configuration (as in config manager) too. I really think we should replace all `*configuration* mentions to *update*/*dispatch update*.

Also I found the example a bit confusing, which ideally it shouldn't be the case, the example should help make things less confusing :P

Marenz commented:
I don't like `update` much in this context either though. How about `DispatchParameterUpdate`

llucax commented:
> I don't like `update` much in this context either though. How about `DispatchParameterUpdate`

Why don't you like `update`? Adding `Parameter` doesn't make it better IMHO, there is no concept of a dispatch parameter in the API. Another option would be `change` instead of `update`, but for me `update` sounds more accurate in this context, this is why I wonder what you don't like about `update`.

Marenz committed:
Fix wrong running behaviors with duration=None

* Dispatch.running() was always returning RUNNING for duration None
* Dispatch.next_run_after() was made aware that there is no next run for
  inf. Duration
* Dispatch._until() now raises when used with Dispatch.None
* Actor: delay marking a Dispatch as deleted, otherwise checks wrongly
  detect it as no longer running.

Signed-off-by: Mathias L. Baumann <mathias.baumann@frequenz.com>

llucax commented on RELEASE_NOTES.md:
```diff
@@ -10,7 +10,7 @@
 ## New Features
 
-<!-- Here goes the main new features and examples or instructions on how to use them -->
+* We now provide the `DispatchRunnerActor` class, a class to manage actors based on incoming dispatches.
```
Given this description, `DispatchManagingActor` doesn't sound like a bad option either. But again, bike shedding, I think it is only marginally better.

llucax reviewed (commented):
I see there was not much progress or replies on my comments. I guess you are focusing on getting shit to work, which sounds totally reasonable, but just saying that I didn't review much in this round.

Marenz committed:
Rewrite internal architecture to be more flexible

The internal scheduling architecture was rewritten:

Instead of using one task per dispatch that will start/stop them,
we now have a queue of events and a timer that is rescheduled
to the earliest next event each time it triggers.

Previously we had to do complicated re-scheduling and cancelling of
tasks, now we always just update the event in the queue.

Signed-off-by: Mathias L. Baumann <mathias.baumann@frequenz.com>

llucax commented on src/frequenz/dispatch/_actor_runner.py:
```diff
@@ -0,0 +1,184 @@
+    """Additional payload."""
+
+
+class DispatchRunnerActor(Actor):
```
bikeshedding: I'm more and more convinced this is more a `DispatchManagingActor`.

llucax commented on src/frequenz/dispatch/_actor_runner.py:
```diff
@@ -0,0 +1,184 @@
+    dry_run: bool
+    """Whether this is a dry run."""
+
+    payload: dict[str, Any]
```
bikeshedding: Payload sounds weird to me. For a whole dispatch it makes sense (or at least it was the best we could find) but in this context maybe something like `options` is more intuitive? To be honest I think `options` might be a good choice for the specs too.

Marenz replied:
how about `inputs`?

llucax replied:
Yeah, that could work too, I think we discussed `options` at some point but wasn't perfectly accurate as some "options" might not be *optional*. I guess `inputs` is an improvement in that sense. Maybe `parameters` could work too.

llucax commented on src/frequenz/dispatch/_actor_runner.py:
```diff
@@ -0,0 +1,184 @@
+
+@dataclass(frozen=True, kw_only=True)
+class DispatchUpdate:
+    """Event emitted when the dispatch configuration changes."""
```
Following my quest to remove "config" from dispatch...

```suggestion
    """Event emitted when the dispatch changes."""
```

llucax commented on src/frequenz/dispatch/_actor_runner.py:
```diff
@@ -0,0 +1,184 @@
+
+                self.set_components(config.components)
+                self._dry_run = config.dry_run
+                self._payload = config.payload
```
```suggestion
        def __init__(self, updates_channel: Receiver[DispatchUpdate]):
            super().__init__()
            self._updates_channel = updates_channel
            self._dry_run: bool
            self._payload: dict[str, Any]

        async def _run(self) -> None:
            while True:
                update = await self._updates_channel.receive()
                print("Received update:", update)

                self.set_components(update.components)
                self._dry_run = update.dry_run
                self._payload = update.payload
```

llucax commented on src/frequenz/dispatch/_actor_runner.py:
```diff
@@ -0,0 +1,184 @@
+
+        async def _run(self) -> None:
+            while True:
+                config = await self._config_channel.receive()
```
And again:

```py
        async for dispatch in self._dispatch_rx:
            await self._handle_dispatch(dispatch=dispatch)
```

But actually this example doesn't look very realistic, I guess the actor should do something else than just *listening* to updates. Maybe it is worth adding some other logic to represent that, maybe implement a `ClockActor` that ticks every X time, where X is a dispatch option? So we have a `select()` instead with a `timer` (printing the time) and the `updates_channel`.

llucax replied:
* #56

llucax commented on src/frequenz/dispatch/_actor_runner.py:
```diff
@@ -0,0 +1,184 @@
+            key=key
+        )
+
+        # Create config channel to receive (re-)configuration events pre-start and mid-run
```
```suggestion
        # Create updates channel to receive update events pre-start and mid-run
```

llucax commented on src/frequenz/dispatch/_actor_runner.py:
```diff
@@ -0,0 +1,184 @@
+        # Create config channel to receive (re-)configuration events pre-start and mid-run
+        dispatch_updates_channel = Broadcast[DispatchUpdate](name="dispatch_updates_channel")
+
+        # Start actor and supporting actor, give each a config channel receiver
```
```suggestion
        # Start actor and supporting actor, give each a updates channel receiver
```

llucax commented on src/frequenz/dispatch/_actor_runner.py:
```diff
@@ -0,0 +1,184 @@
+            actors=frozenset([my_actor, supporting_actor]),
+            dispatch_type="EXAMPLE",
+            running_status_receiver=status_receiver,
+            configuration_sender=dispatch_updates_channel.new_sender(),
```
```suggestion
            updates_sender=dispatch_updates_channel.new_sender(),
```

llucax commented on src/frequenz/dispatch/_actor_runner.py:
```diff
@@ -0,0 +1,184 @@
+        actors: frozenset[Actor],
+        dispatch_type: str,
+        running_status_receiver: Receiver[Dispatch],
+        configuration_sender: Sender[DispatchUpdate] | None = None,
```
```suggestion
        updates_sender: Sender[DispatchUpdate] | None = None,
```

llucax commented on src/frequenz/dispatch/_actor_runner.py:
```diff
@@ -0,0 +1,184 @@
+            actors: The actors to handle.
+            dispatch_type: The type of dispatches to handle.
+            running_status_receiver: The receiver for dispatch running status changes.
+            configuration_sender: The sender for dispatch configuration events
```
```suggestion
            updates_sender: The sender for dispatch updates.
```

llucax commented on src/frequenz/dispatch/_actor_runner.py:
```diff
@@ -0,0 +1,184 @@
+        self._dispatch_rx = running_status_receiver
+        self._actors = actors
+        self._dispatch_type = dispatch_type
+        self._configuration_sender = configuration_sender
```
```suggestion
        self._updates_sender = updates_sender
```

llucax commented on src/frequenz/dispatch/_actor_runner.py:
```diff
@@ -0,0 +1,184 @@
+                await self._stop_actors("Dispatch stopped")
+            case RunningState.RUNNING:
+                if self._configuration_sender is not None:
+                    _logger.info("Updated configuration by dispatch %s", dispatch.id)
```
```suggestion
                if self._updates_sender is not None:
                    _logger.info("Updated by dispatch %s", dispatch.id)
```

llucax commented on src/frequenz/dispatch/_dispatch.py:
```diff
@@ -185,6 +192,7 @@ def next_run_after(self, after: datetime) -> datetime | None:
         if (
             not self.recurrence.frequency
             or self.recurrence.frequency == Frequency.UNSPECIFIED
+            or self.duration is None  # Infinite duration
```
For a different PR (or to convert to an issue in the client repo):

At the python client level we could actually very easily use special values to represent `INFINITE` and `INSTANTANEOUS`, so the code gets more readable (without the need for comments).

We could set the type of `duration: timedelta | Duration` where duration is a enum for example, then we can also use `match` like this:

```py
match dispatch.duration:
    case Duration.INFINITE: ...
    case Duration.INSTANTANEOUS: ...
    case timedelta(): ... (here we know it is > 0)
    case unexpected: assert_never(unexpected)
```

Taking it even a bit further, we could have a `PositiveTimedelta` object that even guarantee the time is positive.

llucax replied:
* https://github.com/frequenz-floss/frequenz-client-dispatch-python/issues/93

llucax commented on src/frequenz/dispatch/actor.py:
```diff
@@ -59,31 +45,97 @@ def __init__(
+                        self._dispatches.pop(dispatch.id)
+                        await self._update_dispatch_schedule_and_notify(None, dispatch)
+
+                        dispatch._set_deleted()  # pylint: disable=protected-access
```
Also for a separate issue: if this function is used externally, then it should be public. If we want to restrict access only to certain users, one trick I've been using is to use either a `Protocol` for the public interface (where you don't expose the `set_deleted()` method) or just plain inheritance (`Dispatch` without `set_delete()` and a `DeletableDispatch` that inherits from it; internally you use `DeletableDispatch` but when returning it to the users, you return a plain `Dispatch`).

This actually made me think why are are storing state in a immutable object? That looks weird.
Another approach, since the `Dispatch`'s `deleted` `bool` and `running_state_change_synced` are only attached as mutable properties for internal **state** tracking, it might make more sense to separate the **state** from the immutable dispatch, and have a `DispatchState` **mutable** class and a separate `dict` to map dispatch ID to its state.

You can pass the state object to the `Dispatch` constructor and store it as a private property and expose *read-only* properties to access the state, if it needs to be accessed externally, and store the `dict` and mutate it from this actor only.

llucax replied:
* #57

llucax commented on tests/test_frequenz_dispatch.py:
```diff
@@ -15,18 +15,30 @@
-def event_loop_policy() -> async_solipsism.EventLoopPolicy:
-    """Return an event loop policy that uses the async solipsism event loop."""
-    return async_solipsism.EventLoopPolicy()
+@fixture(scope="session")
```
:warning:  This shouldn't be necessary and I really think we shouldn't go this way. Things should work if every test runs in its own loop and if it doesn't it means some cleanup is missing or messed up. :warning: 

Oh, I just noticed it also goes back from the correct `event_loop_policy()` to the old (and deprecated) `event_loop()` which is supposed to stop working soon :tm:. This is definitely not an option.

Marenz replied:
I am happy to take any working alternative ;)

llucax reviewed (commented):
Very nice to see the new scheduling algorithm! :tada: Way better than what we had before.

Lots of comments, so here is a summary:

Things I think needs to be addressed before merging:

1. These commits need a commit message with a body that describes what's going on, it is really hard to review them without more information of what was wrong and how it was fixed:

    * `Fix wrong running behaviors with duration=None`
    * `Rewrite internal architecture to be more flexible`
    * `Add dispatch runner` (less important but it would be nice, is not a trivial class)
  
2. Finish the renaming of `config` -> `update`
3. The hack in the tests (session-wide `event_loop` fixture), we can't really merge this if something is so fucked up that makes tests hang or make the interpreter segfault :grimacing: 
4. Figure out if we really need to support multiple actors, it is still make noise in my head, my feeling is we should encourage having one actor to handle one dispatch type, if that actor needs other supporting actors, it should be the actor itself in charge of spawning them.
5. If we are doing more renaming (like `DispatchManagingActor` and `payload` -> `options`), it might be good to do them now too to avoid having breaking changes in the future.

The rest is minor or too big for this PR, so I will create follow-up issues.

Marenz commented:
And I still think "update" sucks and "configuration" was better. "Update" doesn't fit at all in the whole naming scheme

Marenz commented:
Addressed points 1 - 5

llucax commented on src/frequenz/dispatch/_managing_actor.py:
```diff
@@ -0,0 +1,178 @@
+            case RunningState.STOPPED:
+                _logger.info("Stopped by dispatch %s", dispatch.id)
+                await self._stop_actor("Dispatch stopped")
+            case RunningState.RUNNING:
+                if self._updates_sender is not None:
+                    _logger.info("Updated by dispatch %s", dispatch.id)
+                    await self._updates_sender.send(
+                        DispatchUpdate(
+                            components=dispatch.selector,
+                            dry_run=dispatch.dry_run,
+                            options=dispatch.payload,
+                        )
+                    )
+
+                _logger.info("Started by dispatch %s", dispatch.id)
+                self._start_actor()
```
I just realized this is tricky. The actor needs to make sure it reads from the updates channel in the `_run()` method before doing any other work, otherwise it might be running with a missing configuration until the update message arrives. I think we should put this more clearly in the docs, including the example. But we can do it as a followup PR.

llucax replied:
* https://github.com/frequenz-floss/frequenz-dispatch-python/issues/56

llucax commented on src/frequenz/dispatch/_managing_actor.py:
```diff
@@ -0,0 +1,178 @@
+        # Create update channel to receive dispatch update events pre-start and mid-run
+        dispatch_updates_channel = Broadcast[DispatchUpdate](name="dispatch_updates_channel")
+
+        # Start actor and supporting actor, give each a config channel receiver
```
```suggestion
        # Start actor and supporting actor, give each an updates channel receiver
```

llucax commented on tests/test_mananging_actor.py:
```diff
@@ -0,0 +1,164 @@
+    actor: Actor
+    runner_actor: DispatchManagingActor
+    running_status_sender: Sender[Dispatch]
+    configuration_receiver: Receiver[DispatchUpdate]
```
```suggestion
    updates_receiver: Receiver[DispatchUpdate]
```

llucax commented on tests/test_mananging_actor.py:
```diff
@@ -0,0 +1,164 @@
+async def test_env() -> AsyncIterator[TestEnv]:
+    """Create a test environment."""
+    channel = Broadcast[Dispatch](name="dispatch ready test channel")
+    config_channel = Broadcast[DispatchUpdate](name="dispatch config test channel")
```
```suggestion
    config_channel = Broadcast[DispatchUpdate](name="dispatch updates test channel")
```

llucax commented on tests/test_mananging_actor.py:
```diff
@@ -0,0 +1,164 @@
+        actor=actor,
+        dispatch_type="UNIT_TEST",
+        running_status_receiver=channel.new_receiver(),
+        updates_sender=config_channel.new_sender(),
```
```suggestion
        updates_sender=updates_channel.new_sender(),
```

llucax commented on tests/test_mananging_actor.py:
```diff
@@ -0,0 +1,164 @@
+        actor=actor,
+        runner_actor=runner_actor,
+        running_status_sender=channel.new_sender(),
+        configuration_receiver=config_channel.new_receiver(),
```
```suggestion
        updates_receiver=updates_channel.new_receiver(),
```

llucax commented on tests/test_mananging_actor.py:
```diff
@@ -0,0 +1,164 @@
+    await test_env.running_status_sender.send(Dispatch(dispatch))
+    fake_time.shift(timedelta(seconds=1))
+
+    event = await test_env.configuration_receiver.receive()
```
```suggestion
    event = await test_env.updates_receiver.receive()
```

llucax commented on tests/test_mananging_actor.py:
```diff
@@ -0,0 +1,164 @@
+    await test_env.running_status_sender.send(Dispatch(dispatch))
+    fake_time.shift(timedelta(seconds=1))
+
+    event = await test_env.configuration_receiver.receive()
```
```suggestion
    event = await test_env.updates_receiver.receive()
```

llucax reviewed (commented):
A few more instances of *config* and we need support for multiple actors for now (temporarily). I will also create issues for the stuff we can address in the future.

Marenz committed:
DispatchManagingActor: Support starting/stopping of multiple actors

Signed-off-by: Mathias L. Baumann <mathias.baumann@frequenz.com>

Marenz commented:
* Added multiple actors back
* fixed confi -> up stuff

llucax commented on src/frequenz/dispatch/actor.py:
```diff
@@ -59,31 +45,99 @@ def __init__(
+        Interval is chosen arbitrarily, as it will be reset on the first event.
+        """
+
+        self._scheduled_events: list[tuple[datetime, Dispatch]] = []
```
* #58

Marenz committed:
Set dependency to latest dispatch-client

Signed-off-by: Mathias L. Baumann <mathias.baumann@frequenz.com>

llucax reviewed (dismissed):
So this should be ready for merge. @daniel-zullo-frequenz wants to do a release soon using this, so ideally we should have a release too, but I guess for that we first need to release some dependencies too, right?

We need to merge changes in the `client-dispatch` and for that some changes in the `client-base`, right? We had some changes in `channels` too but those were already released in v1.2.0, correct?

Marenz committed:
Update event_loop to explicitly set scope session

As recommended by the deprecation warning

Signed-off-by: Mathias L. Baumann <mathias.baumann@frequenz.com>

Marenz committed:
Wrap scheduler to tolerate somehow invalid dispatches

Signed-off-by: Mathias L. Baumann <mathias.baumann@frequenz.com>

Marenz committed:
Add test that notifications are sent at startup

Signed-off-by: Mathias L. Baumann <mathias.baumann@frequenz.com>

Marenz committed:
Fix wrong Dispatch.running() behavior with duration=None

Signed-off-by: Mathias L. Baumann <mathias.baumann@frequenz.com>

Marenz committed:
DispatchActor: Update timer correctly after schedule modifications

Signed-off-by: Mathias L. Baumann <mathias.baumann@frequenz.com>

Marenz commented:
> [..] we first need to release some dependencies too, right?

Yes.

> We need to merge changes in the client-dispatch and for that some changes in the client-base, right? 

Almost. The changes we merged now actually recreate the streamer object so we don't need any base-client changes.
I have already released the `client-dispatch` version we need now.

> We had some changes in channels too but those were already released in v1.2.0, correct?

Correct. Those are also reflected in the current pyproject.py.


I pushed the last dependency changes and a little new feature so that the actor properly re-reads the dispatches upon disconnect now.

llucax commented on pyproject.toml:
```diff
@@ -165,6 +165,7 @@ disable = [
 [tool.pytest.ini_options]
 testpaths = ["tests", "src"]
 asyncio_mode = "auto"
+asyncio_default_fixture_loop_scope = "function"
```
Maybe we should remove this and keep the default as it is?

Marenz replied:
Then we get a depreciation warning saying the behavior will change and one should set a default

llucax replied:
Oh, OK, I'm confused now. So this has nothing to do with trying to make tests pass and is just a new deprecation message in a new version of `pytest-asyncio`? If so it would be probably nice to put it in a separate commit with a corresponding commit message.

Marenz replied:
I def. tried this to make the test pass, bit it also fixes a deprecation warning, it just didn't help though.

llucax replied:
OK, this seems to be related to an update to https://github.com/pytest-dev/pytest-asyncio/releases/tag/v0.24.0. Issue: https://github.com/pytest-dev/pytest-asyncio/issues/924

llucax commented on tests/test_mananging_actor.py:
```diff
@@ -0,0 +1,164 @@
+from frequenz.dispatch import Dispatch, DispatchManagingActor, DispatchUpdate
+
+
+@fixture(scope="session")
```
Also if it fails even with session, I would leave it to the default too here.

llucax commented on tests/test_frequenz_dispatch.py:
```diff
@@ -15,18 +15,26 @@
 
-# This method replaces the event loop for all tests in the file.
-@fixture
+@fixture(scope="session")
```
Also if it fails even with session, I would leave it to the default too here.

llucax commented:
It looks like now cross-arch tests are consistently failing. This must be definitely a timing issue, the slowest the system is, the more likely we hit this bug and tests hang. Created an issue about it:

* https://github.com/frequenz-floss/frequenz-dispatch-python/issues/66
